src/utils/columnUtils.ts

```diff
@@ -36,22 +41,52 @@ export function getColumnMetrics<R, SR>(metrics: Metrics<R, SR>): ColumnMetrics<
       allocatedWidths += width;
     }
 
-    const column = { ...metricsColumn, width };
+    const column: IntermediateColumn = { ...metricsColumn, width };
+
+    if (rawGroupBy?.includes(column.key)) {
+      column.frozen = true;
+      column.rowGroup = true;
+    }
 
     if (column.frozen) {
       lastFrozenColumnIndex++;
-      columns.splice(lastFrozenColumnIndex, 0, column);
-    } else {
-      columns.push(column);
     }
-  }
+
+    return column;
+  });
+
+  columns.sort(({ key: aKey, frozen: frozenA }, { key: bKey, frozen: frozenB }) => {
+    // Sort select column first:
+    if (aKey === SELECT_COLUMN_KEY) return -1;
+    if (bKey === SELECT_COLUMN_KEY) return 1;
+
+    // Sort grouped columns second, following the groupBy order:
+    if (rawGroupBy?.includes(aKey)) {
+      if (rawGroupBy.includes(bKey)) {
+        return rawGroupBy.indexOf(aKey) - rawGroupBy.indexOf(bKey);
+      }
+      return -1;
+    }
+    if (rawGroupBy?.includes(bKey)) return 1;
+
+    // Sort frozen columns third:
+    if (frozenA) {
+      if (frozenB) return 0;
+      return -1;
+    }
+    if (frozenB) return 1;
+
+    // Sort other columns last:
+    return 0;
+  });
 
   const unallocatedWidth = metrics.viewportWidth - allocatedWidths;
   const unallocatedColumnWidth = Math.max(
     Math.floor(unallocatedWidth / unassignedColumnsCount),
     metrics.minColumnWidth
   );
 
+  const groupBy: string[] = [];
```

Comment:
Should add a comment explaining that this array is a sanitized version of `rawGroupBy`.